addons/knowledge/models/knowledge_article.py

```diff
@@ -0,0 +1,1204 @@
+        """
+        self.ensure_one()
+        if not self.user_can_write:
+            raise AccessError(_("You cannot give access to this article as you are not editor."))
+        share_partner_ids = partner_ids.filtered(lambda partner: partner.partner_share)
+        article = self.sudo()
+        if permission != 'none':
+            article._invite_members(share_partner_ids, 'read', send_mail=send_mail)
+            article._invite_members(partner_ids - share_partner_ids, permission, send_mail=send_mail)
+        else:
+            article._invite_members(partner_ids, permission, send_mail=send_mail)
+        return True
+
+    def _invite_members(self, partner_ids, permission, send_mail=True):
+        """
+        :param partner_ids (Model<res.partner>): Recordset of res.partner
+        :param permission (string): permission ('none', 'read' or 'write')
+        :param send_mail (boolean): Flag indicating whether an email should be sent
+        """
+        self.ensure_one()
+        if not partner_ids:
+            return
+        members = self.article_member_ids.filtered_domain([('partner_id', 'in', partner_ids.ids)])
+        if members:
+            members.write({'permission': permission})
+        partners = partner_ids - members.mapped('partner_id')
+        if not partners:
+            return
+        self.write({
+            'article_member_ids': [(0, 0, {
+                'partner_id': partner.id,
+                'permission': permission
+            }) for partner in partners]
+        })
+        if permission != 'none' and send_mail:
+            for partner in partners:
+                self._send_invite_mail(partner)
+
+    def _send_invite_mail(self, partner):
+        self.ensure_one()
+        subject = _("Invitation to access %s", self.name)
+        partner_lang = get_lang(self.env, lang_code=partner.lang).code
+        body = self.env['ir.qweb'].with_context(lang=partner_lang)._render('knowledge.knowledge_article_mail_invite', {
+            'record': self,
+            'user': self.env.user,
+            'recipient': partner,
+            'link': self._get_invite_url(partner),
+        })
+
+        self.with_context(lang=partner_lang).message_notify(
+            partner_ids=partner.ids, body=body, subject=subject,
+            email_layout_xmlid='mail.mail_notification_light'
+        )
+
+    def _get_invite_url(self, partner):
+        self.ensure_one()
+        return url_join(self.get_base_url(), "/article/%s/invite/%s" % (self.id, partner.id))
+
+    ###########
+    #  Tools
+    ###########
+
+    def _get_internal_permission(self, article_ids=False, check_access=False, check_write=False):
+        """ We don't use domain because we cannot include properly the where clause in the custom sql query.
+        The query's output table and fields names does not match the model we are working on"""
+        domain = []
+        args = []
+        if article_ids:
```

Comment:
zizisse because I use this method in 2 cases: 
- in _search_* methods, where I need to select * the article in the DB.
- in compute methods, where I need to select only the article on self. 

I'm not sure I can do if len(self) == 0 : select * ELSE select where id in article_ids.
I wonder if in some cases, self, inside a _search_* method, is sometime empty, sometimes it contain the "currently active article"
would worth the check